core/src/main/scala/kafka/admin/AdminClient.scala

```diff
@@ -15,38 +15,54 @@ package kafka.admin
 import java.nio.ByteBuffer
 import java.util.{Collections, Properties}
 import java.util.concurrent.atomic.AtomicInteger
+import java.util.concurrent.{TimeUnit, Future}
 
+import kafka.admin.AdminClient.PurgeDataResult
 import org.apache.kafka.common.requests.ApiVersionsResponse.ApiVersion
 import kafka.common.KafkaException
 import kafka.coordinator.GroupOverview
 import kafka.utils.Logging
 import org.apache.kafka.clients._
-import org.apache.kafka.clients.consumer.internals.{ConsumerNetworkClient, ConsumerProtocol, RequestFuture}
+import org.apache.kafka.clients.consumer.internals.{RequestFutureAdapter, ConsumerNetworkClient, ConsumerProtocol, RequestFuture}
 import org.apache.kafka.common.config.ConfigDef.{Importance, Type}
 import org.apache.kafka.common.config.{AbstractConfig, ConfigDef}
 import org.apache.kafka.common.metrics.Metrics
 import org.apache.kafka.common.network.Selector
 import org.apache.kafka.common.protocol.{ApiKeys, Errors}
 import org.apache.kafka.common.requests._
 import org.apache.kafka.common.requests.OffsetFetchResponse
-import org.apache.kafka.common.utils.{Time, Utils}
+import org.apache.kafka.common.utils.{KafkaThread, Time, Utils}
 import org.apache.kafka.common.{Cluster, Node, TopicPartition}
 
 import scala.collection.JavaConverters._
 import scala.util.Try
 
 class AdminClient(val time: Time,
                   val requestTimeoutMs: Int,
+                  val retryBackoffMs: Long,
                   val client: ConsumerNetworkClient,
                   val bootstrapBrokers: List[Node]) extends Logging {
 
+  @volatile var running: Boolean = true
+
+  val networkThread = new KafkaThread("admin-client-network-thread", new Runnable {
+    override def run() {
+      while (running) {
+        client.poll(retryBackoffMs)
+        // Avoid busy loop so that user thread can get the lock on the ConsumerNetworkClient to enqueue requests.
+        if (client.pendingRequestCount() == 0)
+          Thread.sleep(retryBackoffMs)
```

Comment:
When there is no pending requests, `ConsumerNetworkClient.poll(...)` blocks for up to `retryBackoffMs`. It means the thread does poll() in kind of busy looping manner if we don't sleep here. Also, the thread holds the lock on `ConsumerNetworkClient` while blocked on `selector.poll()` which means the user will be blocked waiting for lock in order to enqueue requests.